### PR TITLE
Clean up terminology of derived secrets/keys

### LINKED ISSUE #630: Secrets and Keys
The spec currently uses a mix of terminologies for keys/secrets. In the key schedule we have the following:
------
sender_data_secret | "sender data"
encryption_secret | "encryption"
exporter_secret | "exporter"
authentication_secret | "authentication"
external_secret | "external"
confirmation_key | "confirm"
membership_key | "membership"
resumption_secret | "resumption"
-------

It would be best to provide a coherency to how we use these terms and check for consistency throughout, to avoid confusion for readers:

- A simple solution would be to use "secret" for all cases, i.e., changing membership_key to membership_secret. 

- Another alternative, which may take more refining but could be useful to readers is to separate out terms based on items that are for internal protocol use versus external protocol use. For example, encryption, confirmation, membership, etc., would use "key" and exporter, authentication, etc., could use "secret", or vice versa. 

--Break--

Ref the mailinglist on confirmation vs authentication secrets, perhaps the naming of these two are a little misleading - authentication normally being within bounds of a protocol while confirmation is external to it. This confusion probably sparked some of that discussion to begin with. It _could_ be worth considering switching those terms for clarity.

### PR DESCRIPTION
**Pull request by bifurcation**
Fixes #630

### CODE REVIEW

File: draft-ietf-mls-protocol.md [hunk 1]
```diff
@@ -2470,16 +2470,16 @@ psk_secret (or 0) --> KDF.Extract
...
-| `resumption_secret`     | "resumption"    |
+| Label            | Secret                | Purpose                                                     |
+|:-----------------|:----------------------|:------------------------------------------------------------|
+| "sender data"    | `sender_data_secret`  | Deriving keys to encrypt sender data                        |
+| "encryption"     | `encryption_secret`   | Deriving message encryption keys (via the secret tree)      |
+| "exporter"       | `exporter_secret`     | Deriving exported secrets                                   |
+| "external"       | `external_secret`     | Deriving the external init key                              |
+| "confirm"        | `confirmation_key`    | Computing the confirmation MAC for an epoch                 |
+| "membership"     | `membership_key`      | Computing the membership MAC for an MLSPlaintext            |
+| "resumption"     | `resumption_psk`      | Proving membership in a prior epoch (via a PSK)             |
```
Comment (TWal) [inline draft-ietf-mls-protocol.md:2481]: Isn't it to prove the membership of the current epoch, to be used in a future epoch?

**Review by TWal — COMMENTED** [review]
Things are easier to understand like this!

The "Purpose" column is quite valuable I think, I remember I was quite overwhelmed when I first read it and had no idea what all those secret / keys were used for. This helps to understand the big picture.

**Review by kkohbrok — COMMENTED** [review]
Looks good in general, although I think we should keep the authentication secret a secret rather than a code, since it might be used to do other things than just straight-up compare it. Since it's up to the application to decide how exactly to use it, we should call it a secret and if the application wants to do something with it, it should probably do an additional derivation anyway to make sure there's no potential key-reuse.

File: draft-ietf-mls-protocol.md [hunk 2]
```diff
@@ -2470,16 +2470,16 @@ psk_secret (or 0) --> KDF.Extract
...
-| `resumption_secret`     | "resumption"    |
+| Label            | Secret                | Purpose                                                     |
+|:-----------------|:----------------------|:------------------------------------------------------------|
+| "sender data"    | `sender_data_secret`  | Deriving keys to encrypt sender data                        |
+| "encryption"     | `encryption_secret`   | Deriving message encryption keys (via the secret tree)      |
+| "exporter"       | `exporter_secret`     | Deriving exported secrets                                   |
+| "external"       | `external_secret`     | Deriving the external init key                              |
+| "confirm"        | `confirmation_key`    | Computing the confirmation MAC for an epoch                 |
+| "membership"     | `membership_key`      | Computing the membership MAC for an MLSPlaintext            |
+| "resumption"     | `resumption_psk`      | Proving membership in this epoch (via a PSK injected later) |
```
[the remark this answers is not in the document]
Comment (bifurcation) [reply draft-ietf-mls-protocol.md:2481]: Yeah, it gets used via the `PreSharedKey` proposal, so let's just call it that :)

File: draft-ietf-mls-protocol.md [hunk 3]
```diff
@@ -2470,16 +2470,16 @@ psk_secret (or 0) --> KDF.Extract
...
+| "authentication" | `authentication_code` | Confirming that two clients have the same view of the group |
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply draft-ietf-mls-protocol.md:2482]: Ah, I see what you mean. But the same goes for the exporter secret, doesn't it?

[the remark this answers is not in the document]
Comment (bifurcation) [reply draft-ietf-mls-protocol.md:2482]: Same goes for the export**ed** secrets, and we just call those `exported_value`.  But the `exporter_secret` is an input to the key derivation, so in keeping with the "secret" == "input to HKDF" scheme, it makes sense.

[the remark this answers is not in the document]
Comment (kkohbrok) [reply draft-ietf-mls-protocol.md:2482]: Ah, good point. I had missed that. I guess if we wanted to use the authentication secret/code for something else than comparison, we'd have to go the same route as with the exporter secret/exported values. I'd be ok to call it "code", at least until we figure out a good way to use it beyond comparison.

**Comment by mulmarta** [comment]
Looks good! I would maybe add 1 sentence explaining the rule for what is called a secret and what is called a key. The rule says that a secret will go through some key derivation to obtain a key/keys, is that right?

**Comment by bifurcation** [comment]
@mulmarta Yep, that's exactly right.  Added a sentence.

Thanks for the feedback everyone!

File: draft-ietf-mls-protocol.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkohbrok) [reply draft-ietf-mls-protocol.md:2482]: Then we should really give it the same treatment as the exporter secret and specify in the spec how other secrets can be derived from it. Either that, or merge the two.

**Pull request merged**